File: rust/parity/specs/backup_integrity_checker.py
"""Parity spec for PiBackupIntegrityChecker.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiBackupIntegrityChecker"

_mod = load_py_agent("pi_backup_integrity_checker.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiBackupIntegrityChecker()
    out = agent.check_backup(_mod.BackupInput(**data))
    return out.model_dump()


_CLEAN = "encryption: true\nreplication: true\nretention: 30d"
_UNENCRYPTED = "encryption: false\nreplication: true\nretention: 30d"
_UNENCRYPTED_WORD = "Backups are UNENCRYPTED in the staging vault."
_NO_REPLICATION = "encryption: true\nreplication: disabled\nretention: 30d"
_REPLICATE_FALSE = "encryption: true\nreplicate=false\nretention: 30d"
_SHORT_RETENTION = "encryption: true\nreplication: true\nretention: 1d"
# Mixed case input exercises the .lower() normalisation path.
_MIXED_CASE = "Encryption: FALSE\nReplication: DISABLED\nRetention: 0"

SAMPLES = [
    # Clean / passing config.
    {"input": {"backup_config": _CLEAN}},
    # Unencrypted -> risk 85 -> FAILED_COMPLIANCE under strict default.
    {"input": {"backup_config": _UNENCRYPTED}},
    # "unencrypted" keyword path.
    {"input": {"backup_config": _UNENCRYPTED_WORD}},
    # Single point of failure (replication disabled) -> risk 70.
    {"input": {"backup_config": _NO_REPLICATION}},
    # replicate=false variant -> risk 70.
    {"input": {"backup_config": _REPLICATE_FALSE}},
    # Short retention path -> risk 60.
    {"input": {"backup_config": _SHORT_RETENTION}},
    # All three issues + mixed case -> highest risk wins (85).
    {"input": {"backup_config": _MIXED_CASE}},
    # Empty config -> nothing flagged, PASSED.
    {"input": {"backup_config": ""}},
    # Non-strict env -> risk > 30 but is_secure stays True -> WARN_COMPLIANCE.
    {"input": {"backup_config": _UNENCRYPTED},
     "env": {"PI_BACKUP_STRICT_MODE": "false"}},
    # Strict env explicitly true -> FAILED_COMPLIANCE.
    {"input": {"backup_config": _UNENCRYPTED},
     "env": {"PI_BACKUP_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/backup_integrity_checker.rs
//! Port of `pi_micro_agents/pi_backup_integrity_checker.py`.
//!
//! Verifies multi-region backup replication, recovery checkpoints, and active
//! vault lock policies. Behaviour is a line-for-line mirror of the Python
//! original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub backup_config: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub issues: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: if the env var is set, strict iff its value is
/// (case-insensitively) "true"; if unset, default to strict (true).
fn is_strict_mode() -> bool {
    match std::env::var("PI_BACKUP_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn check_backup(input: &Input) -> Output {
    let content = input.backup_config.to_lowercase();
    let mut issues: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Non-encrypted backups
    if content.contains("encryption: false")
        || content.contains("encryption: disabled")
        || content.contains("unencrypted")
    {
        issues.push(
            "Unencrypted Backups: Backed up assets are stored without standard encryption controls."
                .to_string(),
        );
        risk_score = risk_score.max(85.0);
    }

    // Single region, no replication
    if content.contains("replication: false")
        || content.contains("replicate=false")
        || content.contains("replication: disabled")
    {
        issues.push(
            "Single Point of Failure: No cross-region or multi-zone replication configuration."
                .to_string(),
        );
        risk_score = risk_score.max(70.0);
    }

    // Insecure or missing retention configuration
    if content.contains("retention: 0")
        || content.contains("retention: 1d")
        || content.contains("retention: 1")
    {
        issues.push(
            "Short Retention Period: Backup assets are retained for less than a compliant lifecycle duration."
                .to_string(),
        );
        risk_score = risk_score.max(60.0);
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec {
        "PASSED".to_string()
    } else {
        "FAILED_COMPLIANCE".to_string()
    };
    if risk_score > 0.0 && is_sec {
        status = "WARN_COMPLIANCE".to_string();
    }

    Output {
        is_secure: is_sec,
        issues,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = check_backup(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(cfg: &str) -> Output {
        check_backup(&Input {
            backup_config: cfg.into(),
        })
    }

    #[test]
    fn clean_config_passes() {
        let o = run("encryption: true\nreplication: true\nretention: 30d");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.issues.is_empty());
    }

    #[test]
    fn unencrypted_fails_compliance() {
        let o = run("encryption: false");
        assert!(!o.is_secure);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
        assert_eq!(o.risk_score, 85.0);
        assert_eq!(o.issues.len(), 1);
    }

    #[test]
    fn short_retention_only_warns_when_below_threshold() {
        // retention: 1 -> risk 60.0 (> 30) so it fails under strict mode
        let o = run("retention: 1");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 60.0);
        assert_eq!(o.status, "FAILED_COMPLIANCE");
    }
}
